elasticdl/pkg/ps/server.go

```diff
@@ -2,52 +2,119 @@ package ps
+	var ps Server
+	ps.Param = NewParameter()
+	if opt == "SGD" {
+		ps.Opt = NewSGDOptimizer(lr)
+	}
+	ps.PSID = PSID
+	ps.lock = &sync.Mutex{}
+	return &ps
+}
+
+// PullVariable pulls variable from server
+func (s *Server) PullVariable(ctx context.Context, in *pb.PullVariableRequest) (*pb.PullVariableResponse, error) {
+	var res pb.PullVariableResponse
+	if !s.Param.InitStatus {
+		res.ModelInitStatus = false
+		return &res, nil
+	}
+	res.ModelInitStatus = true
+	res.Model = &pb.Model{}
+	res.Model.Version = s.Param.Version
+	if s.Param.Version > in.CurrentModelVersion {
+		for _, v := range s.Param.NonEmbeddingParam {
+			res.Model.Param = append(res.Model.Param, common.SerializeTensor(v))
+		}
+	}
+	s.Param.InitStatus = true
+	return &res, nil
 }
 
-func (s *psServer) PullEmbeddingVector(ctx context.Context, in *pb.PullEmbeddingVectorRequest) (*pb.Tensor, error) {
-	// TODO: implement the service.
-	return &pb.Tensor{}, nil
+// PullEmbeddingVector pulls embedding vector from server
+func (s *Server) PullEmbeddingVector(ctx context.Context, in *pb.PullEmbeddingVectorRequest) (*pb.Tensor, error) {
+	if in.Ids == nil {
+		return &pb.Tensor{}, nil
+	}
+	table := s.Param.GetEmbeddingParam(in.Name)
+	if table == nil {
+		return &pb.Tensor{}, fmt.Errorf("Request embedding Table %s not found in Param", in.Name)
+	}
+	t := table.GetEmbeddingVectors(in.Ids)
+	return common.SerializeTensor(t), nil
 }
 
-func (s *psServer) PushModel(ctx context.Context, in *pb.Model) (*empty.Empty, error) {
-	// TODO: implement the service.
-	return &empty.Empty{}, nil
+// PushModel pushes model to server
+func (s *Server) PushModel(ctx context.Context, in *pb.Model) (*empty.Empty, error) {
+	if s.Param.InitStatus {
+		return &empty.Empty{}, nil
+	}
+	s.lock.Lock()
+	err := s.Param.InitFromModelPB(in)
+	if err == nil {
+		s.Param.InitStatus = true
+	}
+	s.lock.Unlock()
+	return &empty.Empty{}, err
 }
 
-func (s *psServer) PushEmbeddingInfo(ctx context.Context, in *pb.Model) (*empty.Empty, error) {
-	// TODO: implement the service.
-	return &empty.Empty{}, nil
+// PushEmbeddingInfo pushes embedding info to server
+func (s *Server) PushEmbeddingInfo(ctx context.Context, in *pb.Model) (*empty.Empty, error) {
+	s.lock.Lock()
+	err := s.Param.InitFromModelPB(in)
+	s.lock.Unlock()
+	return &empty.Empty{}, err
 }
 
-func (s *psServer) PushGradient(ctx context.Context, in *pb.PushGradientRequest) (*pb.PushGradientResponse, error) {
-	// TODO: implement the service.
-	return &pb.PushGradientResponse{}, nil
+// PushGradient pushes gradient to server
+func (s *Server) PushGradient(ctx context.Context, in *pb.PushGradientRequest) (*pb.PushGradientResponse, error) {
+	// TODO(qijun) only support async now
+	var res pb.PushGradientResponse
+	var grads []*common.Tensor
+	for _, gradPB := range in.Gradients {
+		grad := common.DeserializeTensorPB(gradPB)
+		grads = append(grads, grad)
+	}
+	err := s.Opt.ApplyGradients(grads, s.Param)
+	s.Param.Version += int32(1)
```

Comment:
I find that you specify the specific type for constant in many places, e.g. `int32(1)`, `int64(1)`. Is this required by Go?

Reply:
Yes, the type of `1` is `int`. And in Go, `int`/`int32`/`int64` are different types. We have to cast it.